### ISSUE TITLE
Scheduled Execution not working

### ISSUE THREAD

**Issue by askmike1** [rundeck/rundeck#6548]
**Describe the bug**
Scheduled executions are not working (running the job manually runs fine)

**My Rundeck detail**

* Rundeck version: 3.2.6 & 3.3.4
* install type: rpm
* OS Name/version: redhat 7.7
* DB Type/version: MariaDB

**To Reproduce**
Set existing job to run every minute or create new job (test job created with a single step in it, nothing else set besides that, the name and the schedule

**Expected behavior**
Scheduled job should run

**Screenshots**
N/A

**Desktop (please complete the following information):**
 - OS: Windows
 - Browser Edge
 - Version 84

**Additional context**
Stacktrace:

```
[2020-10-09T12:30:00,068] ERROR listeners.ExceptionPrinterJobListener - Exception occurred in job:
org.quartz.JobExecutionException: org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: java.lang.RuntimeException: getAuthContextForUserAndRoles: Cannot get AuthContext without user, roles: patam003, null]
        at org.quartz.core.JobRunShell.run(JobRunShell.java:218) [quartz-2.3.2.jar!/:?]
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573) [quartz-2.3.2.jar!/:?]
Caused by: org.quartz.SchedulerException: Job threw an unhandled exception.
        at org.quartz.core.JobRunShell.run(JobRunShell.java:213) [quartz-2.3.2.jar!/:?]
        ... 1 more
Caused by: java.lang.RuntimeException: getAuthContextForUserAndRoles: Cannot get AuthContext without user, roles: patam003, null
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:1.8.0_265]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:1.8.0_265]
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_265]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_265]
        at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:80) ~[groovy-2.5.6.jar!/:2.5.6]
        at org.codehaus.groovy.reflection.CachedConstructor.doConstructorInvoke(CachedConstructor.java:74) ~[groovy-2.5.6.jar!/:2.5.6]
        at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrap.callConstructor(ConstructorSite.java:84) ~[groovy-2.5.6.jar!/:2.5.6]
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:249) ~[groovy-2.5.6.jar!/:2.5.6]
        at rundeck.services.FrameworkService.getAuthContextForUserAndRoles(FrameworkService.groovy:716) ~[classes!/:?]
        at com.dtolabs.rundeck.core.authorization.AuthContextProvider$getAuthContextForUserAndRoles$2.callCurrent(Unknown Source) ~[?:?]
        at rundeck.services.FrameworkService.getAuthContextForUserAndRolesAndProject(FrameworkService.groovy:712) ~[classes!/:?]
        at com.dtolabs.rundeck.core.authorization.AuthContextProvider$getAuthContextForUserAndRolesAndProject$1.call(Unknown Source) ~[?:?]
        at rundeck.quartzjobs.ExecutionJob.initialize(ExecutionJob.groovy:335) ~[classes!/:?]
        at rundeck.quartzjobs.ExecutionJob$initialize.callCurrent(Unknown Source) ~[?:?]
        at rundeck.quartzjobs.ExecutionJob.execute_internal(ExecutionJob.groovy:120) ~[classes!/:?]
        at sun.reflect.GeneratedMethodAccessor1494.invoke(Unknown Source) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_265]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_265]
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101) ~[groovy-2.5.6.jar!/:2.5.6]
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323) ~[groovy-2.5.6.jar!/:2.5.6]
        at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:351) ~[groovy-2.5.6.jar!/:2.5.6]
        at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:64) ~[groovy-2.5.6.jar!/:2.5.6]
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:168) ~[groovy-2.5.6.jar!/:2.5.6]
        at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy:99) ~[classes!/:?]
        at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy) ~[classes!/:?]
        at sun.reflect.GeneratedMethodAccessor1493.invoke(Unknown Source) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_265]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_265]
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101) ~[groovy-2.5.6.jar!/:2.5.6]
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323) ~[groovy-2.5.6.jar!/:2.5.6]
        at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:263) ~[groovy-2.5.6.jar!/:2.5.6]
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041) ~[groovy-2.5.6.jar!/:2.5.6]
        at groovy.lang.Closure.call(Closure.java:405) ~[groovy-2.5.6.jar!/:2.5.6]
        at groovy.lang.Closure.call(Closure.java:399) ~[groovy-2.5.6.jar!/:2.5.6]
        at groovy.lang.Closure.run(Closure.java:486) ~[groovy-2.5.6.jar!/:2.5.6]
        at com.codahale.metrics.Timer.time(Timer.java:137) ~[metrics-core-4.0.7.jar!/:4.0.7]
        at com.codahale.metrics.Timer$time$0.call(Unknown Source) ~[?:?]
        at rundeck.quartzjobs.ExecutionJob.execute(ExecutionJob.groovy:98) ~[classes!/:?]
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202) ~[quartz-2.3.2.jar!/:?]
        ... 1 more
```

scheduled_execution row:

```
# id, version, arg_string, date_created, day_of_month, day_of_week, description, do_nodedispatch, exec_count, execution_enabled, filter, group_path, hour, job_name, last_updated, log_output_threshold, log_output_threshold_action, log_output_threshold_status, loglevel, minute, month, multiple_executions, next_execution, node_exclude, node_exclude_name, node_exclude_os_arch, node_exclude_os_family, node_exclude_os_name, node_exclude_os_version, node_exclude_precedence, node_exclude_tags, node_filter_editable, node_include, node_include_name, node_include_os_arch, node_include_os_family, node_include_os_name, node_include_os_version, node_include_tags, node_keepgoing, node_rank_attribute, node_rank_order_ascending, node_threadcount, nodes_selected_by_default, orchestrator_id, project, retry, retry_delay, schedule_enabled, scheduled, seconds, server_nodeuuid, success_on_empty_node_filter, time_zone, timeout, total_time, rduser, user_role_list, uuid, workflow_id, year, default_tab, node_threadcount_dynamic, notify_avg_duration_threshold, ref_exec_count, filter_exclude, max_multiple_executions, exclude_filter_uncheck, plugin_config
'523', '0', NULL, '2020-10-09 12:28:24', '?', '*', '', '0', '0', '1', NULL, NULL, '*', 'TEST', '2020-10-09 12:28:24', NULL, 'halt', NULL, 'INFO', '*', '*', '0', NULL, NULL, NULL, NULL, NULL, NULL, NULL, '1', NULL, '0', NULL, NULL, NULL, NULL, NULL, NULL, NULL, '0', NULL, '1', '1', '1', NULL, 'rundeck_upgrade_tests', NULL, NULL, '1', '1', '0', '5159ead5-7c92-4265-b48c-aee25e51c2e5', '0', NULL, NULL, '0', 'patam003', '#group1,#group2,group3,user', 'bd1c812d-0783-41f8-b1b6-478f9762b0e0', '245228', '*', 'nodes', NULL, NULL, '0', NULL, NULL, '0', '{\"ExecutionLifecycle\":null}'
```

**Comment by MegaDrive68k** [comment]
Hi @askmike1 

I'm trying to reproduce your issue on 3.2.6 and 3.3.4 without success, can you try with [3.3.4](https://docs.rundeck.com/docs/upgrading/upgrading-to-rundeck-3.3.html)? It seems like this [issue](https://github.com/rundeck/rundeck/issues/2167).

Greetings.

**Comment by askmike1** [comment]
Whoops, sorry I did try this on 3.3.4. It used to work fine but we noticed a while back it was broken (not sure if it broke with one of the upgrades or something else, we don't run many so probably didn't notice right away)

But I can confirm still an issue on 3.3.4 and only for the scheduled feature.

**Comment by askmike1** [comment]
I also find it odd that it is showing 'roles' as being null, but from the db table row I posted, it has roles listed

**Comment by askmike1** [comment]
Just wanted to check if there was a known bug for this? It is occurring in both rundeck environments we have, regardless of the user.

Currently we are on 3.3.5

**Comment by askmike1** [comment]
Actually, I have appeared to find the problem @MegaDrive68k it is an issue if user_role_list starts with a '#'....

Works: `Admin-MyGroup,#MyOtherGroup`
Doesn't Work: `#MyOtherGroup,Admin-MyGroup`

Not sure what the fix is, but that's 100% the problem. Not even sure why it is caring about user_role_list since even if I put bogus groups there it seems to run just fine based on user (as long as it doesn't start with '#')

**Comment by MegaDrive68k** [comment]
Confirmed (on H2 and MySQL),

Steps to reproduce:

1) For testing purposes, add the following ACL focused on group "#user":

```yaml
description: User, all access.
context:
  project: '.*'
for:
  resource:
    - allow: '*'
  adhoc:
    - allow: '*'
  job: 
    - allow: '*'
  node:
    - allow: '*'
by:
  group: '#user'

---

description: User, all access.
context:
  application: 'rundeck'
for:
  resource:
    - allow: '*'
  project:
    - allow: '*'
  project_acl:
    - allow: '*'
  storage:
    - allow: '*' 
by:
  group: '#user'
```

2) Stop the Rundeck service and add a user to the `realm.properties` file with a role which starts with "#" character, in this test is "user".

```
user:user,#user
```

3) Start the Rundeck service and login as "user".

4) Create a new project and upload the following job:

```xml
<joblist>
  <job>
    <defaultTab>nodes</defaultTab>
    <description></description>
    <executionEnabled>true</executionEnabled>
    <id>891c09fc-b76e-42fc-ac92-1372bb675b75</id>
    <loglevel>INFO</loglevel>
    <name>HelloWorld</name>
    <nodeFilterEditable>false</nodeFilterEditable>
    <plugins />
    <schedule>
      <month month='*' />
      <time hour='*' minute='*' seconds='0/10' />
      <weekday day='*' />
      <year year='*' />
    </schedule>
    <scheduleEnabled>true</scheduleEnabled>
    <sequence keepgoing='false' strategy='node-first'>
      <command>
        <exec>echo "hello world"</exec>
      </command>
    </sequence>
    <uuid>891c09fc-b76e-42fc-ac92-1372bb675b75</uuid>
  </job>
</joblist>
```
5) The job can't run on schedule mode but can run manually.

`service.log` output:

```
[2020-11-12T14:38:30,056] ERROR quartzjobs.ExecutionJob - Unable to start Job execution: getAuthContextForUserAndRoles: Cannot get AuthContext without user, roles: user, null
java.lang.RuntimeException: getAuthContextForUserAndRoles: Cannot get AuthContext without user, roles: user, null
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:1.8.0_272]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:1.8.0_272]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_272]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_272]
	at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:80) ~[groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.reflection.CachedConstructor.doConstructorInvoke(CachedConstructor.java:74) ~[groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrap.callConstructor(ConstructorSite.java:84) ~[groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:59) ~[groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:237) ~[groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:249) ~[groovy-2.5.6.jar!/:2.5.6]
	at rundeck.services.FrameworkService.getAuthContextForUserAndRoles(FrameworkService.groovy:813) ~[classes!/:?]
	at com.dtolabs.rundeck.core.authorization.AuthContextProvider$getAuthContextForUserAndRoles$3.callCurrent(Unknown Source) ~[?:?]
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51) ~[groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:156) ~[groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:176) ~[groovy-2.5.6.jar!/:2.5.6]
	at rundeck.services.FrameworkService.getAuthContextForUserAndRolesAndProject(FrameworkService.groovy:809) ~[classes!/:?]
	at com.dtolabs.rundeck.core.authorization.AuthContextProvider$getAuthContextForUserAndRolesAndProject$2.call(Unknown Source) ~[?:?]
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47) ~[groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:115) ~[groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:143) ~[groovy-2.5.6.jar!/:2.5.6]
	at rundeck.quartzjobs.ExecutionJob.initialize(ExecutionJob.groovy:335) ~[classes!/:?]
	at rundeck.quartzjobs.ExecutionJob$initialize.callCurrent(Unknown Source) ~[?:?]
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51) [groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:156) [groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:176) [groovy-2.5.6.jar!/:2.5.6]
	at rundeck.quartzjobs.ExecutionJob.execute_internal(ExecutionJob.groovy:120) [classes!/:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_272]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_272]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_272]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_272]
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101) [groovy-2.5.6.jar!/:2.5.6]
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323) [groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:351) [groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:64) [groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51) [groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:156) [groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:168) [groovy-2.5.6.jar!/:2.5.6]
	at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy:99) [classes!/:?]
	at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy) [classes!/:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_272]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_272]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_272]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_272]
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101) [groovy-2.5.6.jar!/:2.5.6]
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323) [groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:263) [groovy-2.5.6.jar!/:2.5.6]
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041) [groovy-2.5.6.jar!/:2.5.6]
	at groovy.lang.Closure.call(Closure.java:405) [groovy-2.5.6.jar!/:2.5.6]
	at groovy.lang.Closure.call(Closure.java:399) [groovy-2.5.6.jar!/:2.5.6]
	at groovy.lang.Closure.run(Closure.java:486) [groovy-2.5.6.jar!/:2.5.6]
	at com.codahale.metrics.Timer.time(Timer.java:137) [metrics-core-4.0.7.jar!/:4.0.7]
	at com.codahale.metrics.Timer$time$0.call(Unknown Source) [metrics-core-4.0.7.jar!/:4.0.7]
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47) [groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:115) [groovy-2.5.6.jar!/:2.5.6]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:127) [groovy-2.5.6.jar!/:2.5.6]
	at rundeck.quartzjobs.ExecutionJob.execute(ExecutionJob.groovy:98) [classes!/:?]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202) [quartz-2.3.2.jar!/:?]
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573) [quartz-2.3.2.jar!/:?]
```

Meanwhile, the main recommendation isn't to use special characters for users and groups.

Good catch @askmike1!

**Comment by askmike1** [comment]
Thanks! Though the main recommendation cannot be taken in our case (unfortunately) as this is linked to AD which we do not manage (and our 'distribution lists' start with a '#')

**Comment by askmike1** [comment]
Just to follow up on this, will the fix for this be in an upcoming release?

**Comment by askmike1** [comment]
Haven't tested with 3.3.11 yet, but can confirm it is still an issue with 3.3.10

**Comment by rarango9** [comment]
So I was running 3.3.0 and had to downgrade for some testing, but i noticed something interesting on this.

Scheduled jobs that were saved on version 3.3.0 were saving the user roles in the database like `#TeamName,UserName,$Hello`. Notice of lack of brackets in the string that is saved to the database.

When I downgraded I resaved the job without modifying anything and it updated the string in the database to `["#TeamName","UserName","$Hello"]`.

When the user roles were just a comma delimited string the scheduled run was failing with the `Cannot get AuthContext without user, roles` error, but when in JSON format string it succeeds.

@askmike1 Assuming your still running a newer version in your environment, can you check a scheduled job you know is failing and see what is saved in the table for that entry?

DB -> rundeck -> tables -> scheduled_execution, column is `user_role_list`

If the failing job is saved as a comma delimited string, would you mind manually changing the DB entry to a JSON formatted string and let the scheduled job run without updating to see if it now works?

I did the reverse to try and break a working one in version 3.2.3 and it definitely broke the scheduled run.
```
org.quartz.JobExecutionException: org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: java.lang.RuntimeException: getAuthContextForUserAndRoles: Cannot get AuthContext without user, roles: aranr018, null]
```

**Comment by askmike1** [comment]
We actually just upgraded to 3.4.0 and this issue seems to be resolved with that release from our initial testing.

**Comment by rarango9** [comment]
Think I found the bug which was fixed starting 3.4.0-alpha1
[Avoiding error when parsing user roles that contains commas in it.](https://github.com/rundeck/rundeck/commit/968fe04f5810f27f37a7a0f16372a3c69d2da1f9#diff-5637b327b0a872eb22270f47d22981a35ba497caf51eb659154906028f2ca08fL203)

Did a small test and results line up with what I was seeing in the comment before.

```
import groovy.json.JsonOutput

List userRoleList = ['#TeamName', 'UserName', '$RoleName']

println userRoleList
println ''
println userRoleList?.join(",")
println JsonOutput.toJson(userRoleList)
```

Output
```
[#TeamName, UserName, $RoleName]

#TeamName,UserName,$RoleName
["#TeamName","UserName","$RoleName"]
```

**Comment by MegaDrive68k** [comment]
Fixed on 3.4.X (tested).
